iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/plan/AbstractFragmentParallelPlanner.java

```diff
@@ -133,13 +132,29 @@ protected TDataNodeLocation selectTargetDataNode(TRegionReplicaSet regionReplica
     if (regionReplicaSet.getDataNodeLocationsSize() != availableDataNodes.size()) {
       LOGGER.info("available replicas: {}", availableDataNodes);
     }
+    int targetIndex = getTargetIndex(availableDataNodes);
+    return availableDataNodes.get(targetIndex);
+  }
+
+  private int getTargetIndex(List<TDataNodeLocation> availableDataNodes) {
     int targetIndex;
-    if (!selectRandomDataNode || queryContext.getSession() == null) {
+    if (ReadConsistencyLevel.STRONG == this.readConsistencyLevel
+        || queryContext.getSession() == null) {
       targetIndex = 0;
-    } else {
+    } else if (ReadConsistencyLevel.WEAK == this.readConsistencyLevel) {
       targetIndex = (int) (queryContext.getSession().getSessionId() % availableDataNodes.size());
+    } else if (ReadConsistencyLevel.FOLLOWER_READ == this.readConsistencyLevel) {
+      // The first available data node is always leader which is guaranteed by ConfigNode and
+      // PartitionFetcher in DataNode
+      // We only need to randomly choose any one from [1, availableDataNodes.size()).
+      // SessionId is a unchanged long value for each connection, so we can use that as random seed
+      targetIndex =
```

Comment:
When we only have one replica, maybe we should return 0, return a number >= 1 will cause an error?